bayem/vba.py

```diff
@@ -0,0 +1,519 @@
+
+    def stop_criteria(self, f_new, i_iter):
+        self.n_trials += 1
+
+        if f_new > self.f_old:
+            self.n_trials = 0
+
+        # Update free energy here such that the "stop_criteria" is testable
+        # individually:
+        self.result.f_max = max(self.result.f_max, f_new)
+
+        # stop?
+        if self.n_trials >= self.options.maxtrials:
+            self.result.message = "Stopping because free energy did not "
+            self.result.message += (
+                f"increase within {self.options.maxtrials} iterations."
+            )
+            return True
+
+        if i_iter >= self.options.maxiter:
+            self.result.message = "Stopping because the maximum number of "
+            self.result.message = "iterations is reached."
+            return True
+
+        if abs(f_new - self.f_old) <= self.options.tolerance:
+            self.result.message = "Tolerance reached!"
+            self.result.success = True
+            return True
+
+        # go on.
+        self.f_old = f_new
+        return False
+
+
+
+
+class VBResult:
+    """
+    Somehow inspired by
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.OptimizeResult.html
+    """
+
+    def __init__(self):
+        self.param = None
+        self.noise = {}
+        self.success = False
+        self.message = ""
+        self.free_energies = []
+        self.means = []
+        self.sds = []
+        self.shapes = []
+        self.scales = []
+        self.f_max = -np.inf
+        self.nit = 0
+        self.t = None
+
+    def __str__(self):
+        s = "### VB Result ###\n"
+
+        for member in [
+            "success",
+            "message",
+            "param",
+            "noise",
+            "free_energies",
+            "f_max",
+            "nit",
+            "t",
+        ]:
+            to_print = self.__dict__[member]
+            v = str(to_print).replace("\n", "\n" + " " * 17)
+            s += f"  {member+':':14s} {v}\n"
+
+        return s
+
+    def try_update(self, f_new, mean, precision, shapes, scales, names):
+        self.free_energies.append(f_new)
+        self.means.append(mean)
+        self.sds.append(MVN(mean, precision).std_diag)
+        self.shapes.append(shapes)
+        self.scales.append(scales)
+        if f_new > self.f_max:
+            # update
+            self.f_max = f_new
+            self.param = MVN(mean, precision, name="MVN posterior", names=names)
+
+            for n in shapes:
+                self.noise[n] = Gamma(shape=shapes[n], scale=scales[n])
+
+    def summary(
+        self,
+        gamma_as_sd=False,
+        printer=None,
+        quantiles=[0.05, 0.25, 0.75, 0.95],
+        **tabulate_kwargs,
+    ):
+        if printer is None:
+            printer = print
+
+        data = []
+        p = self.param
+        for i in range(len(p)):
+            dist = p.dist(i)
+            entry = [p.names[i], dist.median(), dist.mean(), dist.std()]
```

Comment:
Regarding the discussion related to the ill-posed problem with C_beam and C_connection, I would argue that the std are not necessarily the interesting outputs, but maybe the eigenvalues/vectors of the posterior.

Reply:
Good point. However this method is inspired by the `summary` methods of other Bayesian inference frameworks, where I have not yet seen such your proposal. That could be something for the expert user to implement...